parser/src/layout.rs

```diff
@@ -0,0 +1,404 @@
+#[derive(Copy, Clone, Debug, PartialEq)]
+enum Context {
+    /// Context which contains several expressions/declarations separated by semicolons
+    Block { emit_semi: bool },
+    /// After brace token
+    Brace,
+    /// After bracket token
+    Bracket,
+    /// After paren token
+    Paren,
+    /// A simple expression
+    Expr,
+    /// In a let expression
+    Let,
+    /// In a type expression
+    Type,
+    /// In an if expression
+    If,
+    /// In a match clause
+    MatchClause,
+    /// In a lambda function
+    Lambda,
+}
+
+struct Contexts {
+    stack: Vec<Offside>,
+}
+
+impl Contexts {
+    fn new() -> Contexts {
+        Contexts { stack: Vec::new() }
+    }
+
+    fn last(&self) -> Option<&Offside> {
+        self.stack.last()
+    }
+
+    fn last_mut(&mut self) -> Option<&mut Offside> {
+        self.stack.last_mut()
+    }
+
+    fn pop(&mut self) -> Option<Offside> {
+        self.stack.pop()
+    }
+
+    fn push(&mut self, offside: Offside) -> Result<(), CombineError<String, String>> {
+        self.check_unindentation_limit(offside)?;
+        self.stack.push(offside);
+        Ok(())
+    }
+
+    fn check_unindentation_limit(&mut self,
+                                 offside: Offside)
+                                 -> Result<(), CombineError<String, String>> {
+        let mut skip_block = false;
+        for other_offside in self.stack.iter().rev() {
+            match other_offside.context {
+                Context::Lambda => {
+                    skip_block = true;
+                    continue;
+                }
+                Context::Block { .. } if skip_block => continue,
+                Context::Brace | Context::Bracket | Context::Paren => return Ok(()),
+                // New context should not be unindented past the closest enclosing block context
+                Context::MatchClause |
+                Context::Type |
+                Context::Let |
+                Context::Block { .. } if offside.location.column <
+                                         other_offside.location.column => (),
+                _ => continue,
+            }
+            debug!("Unindentation error: {:?} < {:?}", offside, other_offside);
+            return Err(CombineError::Message("Line was unindented to far".into()));
+        }
+        Ok(())
+    }
+}
+
+pub struct Layout<'input, Tokens> {
+    tokens: Tokens,
+    unprocessed_tokens: Vec<SpannedToken<'input>>,
+    indent_levels: Contexts,
+}
+
+impl<'input, Tokens> Layout<'input, Tokens>
+    where Tokens: Iterator<Item = SpannedToken<'input>>,
+{
+    pub fn new(tokens: Tokens) -> Layout<'input, Tokens> {
+        Layout {
+            tokens: tokens,
+            unprocessed_tokens: Vec::new(),
+            indent_levels: Contexts::new(),
+        }
+    }
+
+    fn next_token(&mut self) -> SpannedToken<'input> {
+        self.unprocessed_tokens.pop().unwrap_or_else(|| {
+            self.tokens.next().unwrap_or_else(|| {
+                // Blegh
```

Comment:
Might not be a problem (other than being ugly) as the span should not be used for the EOF token anyway, add a comment on that maybe if its true.